Issue 51524: Allowlist for acceptable types of file upload - automation test

labkey-sweta opened the pull request:
#### Rationale
Add automated tests for the Allowed Files list.

These tests will cover:

- Validating the admin page for the Allowed Files List (http://localhost:8080/admin-allowList.view?type=FileExtension).
- Test using the Files web part.
- Test using Sample Types in LKS.
- Test using Lists (tests attachments).
- Test using Messages.

Has coverage for:
[Issue 53039](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53039): Updating an allowed file extension should set a dirty bit on the page.
[Issue 53027](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53027): Cannot insert a row into a sample type if Allowed File Extensions are set and no file is provided.

TeamCity runs are [here](https://teamcity.labkey.org/buildConfiguration/bt20?branch=allowFileExtensionAutoTest&buildTypeTab=overview).

#### Related Pull Requests
- https://github.com/LabKey/limsModules/pull/1409
- https://github.com/LabKey/platform/pull/6677

#### Changes
- Add a AllowedFileExtensionAdminPage.
- Update ShowAdminPage, BaseUpdatePage, FileBrowserHelper
- Add AllowedFileExtensionAdminTest

labkey-sweta committed:
Issue 51524: Allowlist for acceptable types of file upload - automated tests

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Reworking some of the allowed file extension tests.

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Finish tests for SampleTypes and Lists.
Add test to check for adding invalid extensions.

labkey-danield committed:
Add tests for Messages.
Update BaseUpdatePage to have a getter for the title and body.

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Change test workflow to accommodate resolution of issue 53026 (not going to check if fields remain populated).

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Move deleteAllAllowedFileExtensions out of AllowedFileExtensionAdminTest and into BaseWebDriverTest.
Add AllowedFileExtensionsTest in lims/biologics, add testAllowedFilesInSampleTypes.
Add TestScrubber.clearListOfAllowedFileExtensions.

labkey-danield committed:
Add check for dirty bit behavior to AllowedFileExtensionAdminTest
Add AllowedFileExtensionsTest.testAllowedFilesInWorkflow

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-jeckels commented on src/org/labkey/test/BaseWebDriverTest.java:
```diff
@@ -2507,6 +2507,15 @@ protected void reloadStudyFromZip(File folderArchive, boolean validateQueries, i
         waitForPipelineJobsToComplete(pipelineJobs, "Study Reload", expectError);
     }
 
+    protected void deleteAllAllowedFileExtension() throws IOException, CommandException
```
```suggestion
    /** Clears any file extensions that are set as the only allowable names, letting users upload any filename they like */
    protected void deleteAllAllowedFileExtension() throws IOException, CommandException
```

labkey-danield replied:
Added.

labkey-jeckels commented on src/org/labkey/test/TestScrubber.java:
```diff
@@ -213,4 +222,13 @@ private void disableFileUploadSetting() throws IOException, CommandException
         }
     }
+
+    private void clearListOfAllowedFileExtensions() throws IOException, CommandException
```
Can we consolidate this with the copy in `BaseWebDriverTest`?

labkey-danield replied:
I've removed this method. The only place this code should now appear is in `AllowedFileExtensionAdminPage.deleteAllAllowedFileExtension`.

labkey-jeckels commented on src/org/labkey/test/tests/core/admin/AllowedFileExtensionAdminTest.java:
```diff
@@ -0,0 +1,618 @@
+                "StoredAmount", amount);
+        sampleTypeHelper.insertRow(fieldMap);
+
+        // Not sure why we record two exceptions.
+        checkExpectedErrors(2);
```
Should no longer be needed

labkey-danield replied:
Removed.

labkey-jeckels commented on src/org/labkey/test/tests/core/admin/AllowedFileExtensionAdminTest.java:
```diff
@@ -0,0 +1,618 @@
+
+        page.submit();
+
+        // Not sure why we record two exceptions.
+        checkExpectedErrors(2);
```
Should no longer be needed

labkey-danield replied:
Also gone.

labkey-tchad commented on src/org/labkey/test/BaseWebDriverTest.java:
```diff
@@ -2507,6 +2507,15 @@ protected void reloadStudyFromZip(File folderArchive, boolean validateQueries, i
         waitForPipelineJobsToComplete(pipelineJobs, "Study Reload", expectError);
     }
 
+    protected void deleteAllAllowedFileExtension() throws IOException, CommandException
+    {
+        SimplePostCommand command = new SimplePostCommand("admin", "deleteAllValues");
+        Map<String, Object> params = new HashMap<>();
+        params.put("type", "FileExtension");
+        command.setParameters(params);
+        command.execute(createDefaultConnection(), "/");
+    }
```
This method should go in `LabKeySiteWrapper`. That would allow `TestScrubber` to access it.
Alternatively, you could make it a static method in `AllowedFileExtensionAdminPage` (with a `Connection` parameter).

labkey-danield replied:
I've moved it AllowedFileExtensionsAdminPage and made it static. I think it might be more discoverable there.

labkey-tchad commented on src/org/labkey/test/pages/core/admin/AllowedFileExtensionAdminPage.java:
```diff
@@ -0,0 +1,189 @@
+
+    public static AllowedFileExtensionAdminPage beginAt(WebDriverWrapper webDriverWrapper)
+    {
+        webDriverWrapper.beginAt(WebTestHelper.buildURL("admin", "allowList"));
```
The allowList page has several flavors, need to specify the `FileExtension` version.
```suggestion
        webDriverWrapper.beginAt(WebTestHelper.buildURL("admin", "allowList", Map.of("type", "FileExtension")));
```

labkey-danield replied:
Added.

labkey-tchad commented on src/org/labkey/test/pages/core/admin/AllowedFileExtensionAdminPage.java:
```diff
@@ -0,0 +1,189 @@
+        return this;
+    }
+
+    public String clickSaveExpectingError()
+    {
+        String errorText;
+        elementCache().saveExtension.click();
+        errorText = waitForElement(Locators.labkeyError).getText();
+        clearCache();
+        return errorText;
+    }
+
+    public AllowedFileExtensionAdminPage clickSaveExtension()
+    {
+        elementCache().saveExtension.click();
+        clearCache();
+        return this;
+    }
```
All of the buttons (delete, save, and update) trigger a page load, regardless of whether there is an error. These should all wait for a page load and error-free versions should verify that there is no error.
```suggestion
    private AllowedFileExtensionAdminPage clickButtonNoError(WebElement button)
    {
        clickAndWait(button);
        clearCache();
        assertNoLabKeyErrors();
        return this;
    }

    private String clickButtonExpectingError(WebElement button)
    {
        clickAndWait(button);
        clearCache();
        return waitForElement(Locators.labkeyError).getText();
    }

    public String clickSaveExpectingError()
    {
        return clickButtonExpectingError(elementCache().saveExtension);
    }

    public AllowedFileExtensionAdminPage clickSaveExtension()
    {
        return clickButtonNoError(elementCache().saveExtension);
    }
```

And so on for update and delete.

labkey-danield replied:
Updated.

labkey-tchad commented on src/org/labkey/test/pages/core/admin/AllowedFileExtensionAdminPage.java:
```diff
@@ -0,0 +1,189 @@
+
+    public AllowedFileExtensionAdminPage deleteAllExtensions(boolean acceptAlert)
+    {
+        elementCache().deleteAll.click();
+
+        if (acceptAlert)
+        {
+            acceptAlert();
+            shortWait().withMessage("'Delete All' button should have gone away.")
+                    .until(ExpectedConditions.stalenessOf(elementCache().deleteAll));
+        }
+        else
+        {
+            cancelAlert();
+        }
+
+        clearCache();
+        return this;
```
This staleness check will always pass because there is a page load. Need to clear the cache and check the button's visibility.
```suggestion
        if (acceptAlert)
        {
            doAndWaitForPageLoad(() -> {
                elementCache().deleteAll.click();
                acceptAlert();
            });
            clearCache();
            assertFalse("Delete All button should not be present after deleting all extensions", elementCache().deleteAll.isDisplayed());
        }
        else
        {
            elementCache().deleteAll.click();
            cancelAlert();
            assertTrue("Delete All button should be present", elementCache().deleteAll.isDisplayed());
        }

        return this;
```

labkey-danield replied:
Updated.

labkey-tchad commented on src/org/labkey/test/tests/core/admin/AllowedFileExtensionAdminTest.java:
```diff
@@ -0,0 +1,618 @@
+        _containerHelper.createProject(getProjectName(), null);
+
+        goToProjectHome();
+        new PortalHelper(getDriver()).addWebPart("Files");
+        new PortalHelper(getDriver()).addWebPart("Sample Types");
+        new PortalHelper(getDriver()).addWebPart("Lists");
+        new PortalHelper(getDriver()).addWebPart("Messages List");
```
Each `addWebPart` enters and exits page admin mode. Wrap in `doInAdminMode` to avoid all the extra navigations.
```suggestion
        new PortalHelper(driver).doInAdminMode(ph -> {
            ph.addWebPart("Files");
            ph.addWebPart("Sample Types");
            ph.addWebPart("Lists");
            ph.addWebPart("Messages List");
        });
```

labkey-danield replied:
Done.

labkey-tchad commented on src/org/labkey/test/tests/core/admin/AllowedFileExtensionAdminTest.java:
```diff
@@ -0,0 +1,618 @@
+        log(String.format("Verify file with the old extension '%s' is excluded.", oldExtension));
+        uploadToFileWebPartExcluded(excludedTypes);
+
+        allowedFileExtensionAdminPage = goToAdminConsole().clickAllowedFileExtensions();
+        allowedFileExtensionAdminPage.setExtension("not .an extension");
+        String expectedValue = "File extension must start with a '.'";
+        String actualValue = allowedFileExtensionAdminPage.clickSaveExpectingError();
+        checker().withScreenshot()
+                .verifyEquals("Incorrect error message for invalid extension.",
+                        expectedValue, actualValue);
+
+        allowedFileExtensionAdminPage.setExtension(allowedTypes.get(0));
+        expectedValue = String.format("'%s' already exists. Duplicate values not allowed.", allowedTypes.get(0));
+        actualValue = allowedFileExtensionAdminPage.clickSaveExpectingError();
+        checker().withScreenshot()
+                .verifyEquals("Incorrect error message for duplicate extension.",
+                        expectedValue, actualValue);
+
+        allowedFileExtensionAdminPage.setExtension(allowedTypes.get(1).toUpperCase());
+        expectedValue = String.format("'%s' already exists. Duplicate values not allowed.", allowedTypes.get(1).toUpperCase());
+        actualValue = allowedFileExtensionAdminPage.clickSaveExpectingError();
+        checker().withScreenshot()
+                .verifyEquals("Incorrect error message for duplicate extension but different case.",
+                        expectedValue, actualValue);
+
+        allowedFileExtensionAdminPage.setExtension("");
+        expectedValue = "File extension must not be blank.";
+        actualValue = allowedFileExtensionAdminPage.clickSaveExpectingError();
+        checker().withScreenshot()
+                .verifyEquals("Incorrect error message for blank extension value.",
+                        expectedValue, actualValue);
```
Consider moving the admin page validation to a separate test case (duplicate extensions, updating extensions, invalid extensions, delete all cancelling, etc.)

labkey-danield replied:
I have moved most of the tests cases that check the admin page functionality into a separate class:
- Check that allowed file extensions are listed in alphabetical order.
- Canceling  out of 'Delete All'.
- Validate extension value must start with a '.'
- Validate duplicate extensions are not allowed.
- Validate blank extension type is not allowed.

I did leave in test cases that validate actions on the admin page behave as expected in the project:
- Remove an allowed file type and validate files of that type cannot be uploaded.
- Click 'Delete All' and validate any file type can be uploaded.
- Edit an allowed extension, .xls to .xlsx, and validate .xlsx files can be uploaded, but .xls cannot.

As a side note it is not possible to test for dirty bit in LKS with selenium. See [issue 38785](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=38785).

labkey-tchad commented on src/org/labkey/test/tests/core/admin/AllowedFileExtensionAdminTest.java:
```diff
@@ -0,0 +1,618 @@
+        String attachmentField = "Attachment Field";
+
+        log(String.format("Create a list '%s' with attachment field and auto-increment key.", listName));
+        _listHelper.createList(getProjectName(), listName, keyField,
+                new FieldDefinition(attachmentField, FieldDefinition.ColumnType.Attachment));
```
Create list via API.
```suggestion
        new IntListDefinition(listName, keyField)
            .addField(new FieldDefinition(attachmentField, FieldDefinition.ColumnType.Attachment))
            .create(getProjectName(), createDefaultConnection());
```

labkey-danield replied:
Done

labkey-tchad commented on src/org/labkey/test/tests/core/admin/AllowedFileExtensionAdminTest.java:
```diff
@@ -0,0 +1,618 @@
+        // Same as Issue 53026, the fields are cleared after hitting back button. Covewrage for that issue is in
+        // testAllowedFileExtensionsInSampleType test.
+        log("Clear the file field.");
+        FileInput el = FileInput.FileInput(Locator.name("quf_" + attachmentField), getDriver()).findWhenNeeded();
+        executeScript("arguments[0].value = '';", el.getComponentElement());
```
Does `WebElement.clear` work for this?
```suggestion
        FileInput el = FileInput.FileInput(Locator.name("quf_" + attachmentField), getDriver()).findWhenNeeded();
        el.getComponentElement().clear();
```

If it does, consider adding a `clear` method to `FileInput`.

labkey-danield replied:
It appears to. I've added `FileInput.clear`.

labkey-tchad reviewed (changes requested):
The instructions on the admin page say that ".gz" won't allow ".tar.gz" files; should we verify that is true?

labkey-jeckels commented:
> The instructions on the admin page say that ".gz" won't allow ".tar.gz" files; should we verify that is true?

This is a good idea - we should cover a "double extension" like this.

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-chrisj reviewed (commented):
I don't have anything pressing to add, looks good to me tho

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Add a tar.gz file to the mix.
Move deleteAllAllowedFileExtensions to the AllowedFileExtensionAdminPage
Consolidate button code in AllowedFileExtensionAdminPage
Remove calls to checkExpectedErrors.
Adding FileInput.clear method.
Many other code review feedback as well.

labkey-danield committed:
Add a tar.gz file to the mix.
Move deleteAllAllowedFileExtensions to the AllowedFileExtensionAdminPage
Consolidate button code in AllowedFileExtensionAdminPage
Remove calls to checkExpectedErrors.
Adding FileInput.clear method.
Many other code review feedback as well.

labkey-danield commented:
> > The instructions on the admin page say that ".gz" won't allow ".tar.gz" files; should we verify that is true?
> 
> This is a good idea - we should cover a "double extension" like this.

I've added a tar.gz file to the mix.

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Move most of the Admin Page specific test cases into a separate class (AllowedFileExtensionsPageTest).
Add a base class that has some shared code.

labkey-danield committed:
Change test names.

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Commented out the parts of the test that check for dirty bit.
Cleaned up the biologics AllowedFileExtensionsTest (removed reset errors, used existing methods, updated expected error message).
Added AppRegisterCellLinePage.setColumnValue that takes a file parameter.
Added AppRegisterCellLinePage.clickSubmitExpectingError.
Updated ThreadAttachment.clickRemove to return a confirmation dialog.
Added ThreadEditor.clickCreateBtnExpectingError.

labkey-danield committed:
Merge branch 'develop' into fb_allowFileExtensionAutoTest

labkey-danield committed:
Update tests to validate the complete error message.
Fixed some typos in comments.
